nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/PublishGCPubSub.java

```diff
@@ -213,61 +289,139 @@ public List<ConfigVerificationResult> verify(final ProcessContext context, final
+            onTriggerFlowFileStrategy(context, session, stopWatch, flowFileBatch);
+        } else if (ContentInputStrategy.RECORD_ORIENTED.equals(inputStrategy)) {
+            onTriggerRecordStrategy(context, session, stopWatch, flowFileBatch);
+        } else {
+            throw new IllegalStateException(inputStrategy.getValue());
         }
+    }
+
+    private void onTriggerFlowFileStrategy(
+            final ProcessContext context,
+            final ProcessSession session,
+            final StopWatch stopWatch,
+            final List<FlowFile> flowFileBatch) throws ProcessException {
+        final long maxMessageSize = context.getProperty(MAX_MESSAGE_SIZE).asDataSize(DataUnit.B).longValue();
 
-        final long startNanos = System.nanoTime();
-        final List<FlowFile> successfulFlowFiles = new ArrayList<>();
-        final String topicName = getTopicName(context).toString();
+        final MessageTracker messageTracker = new MessageTracker();
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        for (final FlowFile flowFile : flowFileBatch) {
+            if (flowFile.getSize() > maxMessageSize) {
+                final String message = String.format("FlowFile size %d exceeds MAX_MESSAGE_SIZE", flowFile.getSize());
+                messageTracker.add(new FlowFileResult(flowFile, Collections.emptyList(), new IllegalArgumentException(message)));
+            } else {
+                baos.reset();
+                session.exportTo(flowFile, baos);
+                final ApiFuture<String> future = publishOneMessage(context, flowFile, baos.toByteArray());
+                messageTracker.add(new FlowFileResult(flowFile, Collections.singletonList(future)));
+            }
+        }
+        finishBatch(session, flowFileBatch, stopWatch, messageTracker);
+    }
 
+    private void onTriggerRecordStrategy(
+            final ProcessContext context,
+            final ProcessSession session,
+            final StopWatch stopWatch,
+            final List<FlowFile> flowFileBatch) throws ProcessException {
         try {
-            for (FlowFile flowFile : flowFiles) {
-                try {
-                    final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-                    session.exportTo(flowFile, baos);
-                    final ByteString flowFileContent = ByteString.copyFrom(baos.toByteArray());
-
-                    PubsubMessage message = PubsubMessage.newBuilder().setData(flowFileContent)
-                            .setPublishTime(Timestamp.newBuilder().build())
-                            .putAllAttributes(getDynamicAttributesMap(context, flowFile))
-                            .build();
+            onTriggerRecordStrategyInner(context, session, stopWatch, flowFileBatch);
+        } catch (IOException | SchemaNotFoundException | MalformedRecordException e) {
+            throw new ProcessException(e);

```

Comment:
Exceptions should always include a clarifying message. Recommend adding something thing like, "Record publishing failed".
```suggestion
            throw new ProcessException("Record publishing failed", e);

```